go/enclave/rollupchain/rollup_chain.go

```diff
@@ -393,82 +403,91 @@ func (rc *RollupChain) produceBlockSubmissionResponse(block *types.Block, l2Head
+		return nil, fmt.Errorf("could not handle block. Cause: %w", err)
 	}
 
-	rc.logger.Trace(fmt.Sprintf("Calc block state b_%d: Found: r_%d, ",
-		common.ShortHash(block.Hash()), common.ShortHash(*l2Head)))
-
-	return l2Head, isUpdatedRollupHead, nil
+	return l2Head, nil
 }
 
-func (rc *RollupChain) handleGenesisRollup(block *types.Block, rollupsInBlock []*core.Rollup) (*common.L2RootHash, error) {
+// Determines if this is a pre-genesis L2 block, the genesis L2 block, or a post-genesis L2 block.
+func (rc *RollupChain) getBlockType(rollupsInBlock []*core.Rollup) (*BlockType, error) {
+	preGenesis := PreGenesis
+	genesis := Genesis
+	postGenesis := PostGenesis
+
 	genesisRollup, err := rc.storage.FetchGenesisRollup()
 	if err != nil {
-		// If there is no genesis yet and no rollups have arrived, there is nothing to do
-		if errors.Is(err, errutil.ErrNotFound) {
-			if len(rollupsInBlock) == 0 {
-				return nil, errIsPreGenesis
-			}
-		} else {
+		if !errors.Is(err, errutil.ErrNotFound) {
 			return nil, fmt.Errorf("could not retrieve genesis rollup. Cause: %w", err)
 		}
-	}
 
-	// We've already found the genesis rollup.
-	if genesisRollup != nil {
-		// Re-processing the block that contains the genesis rollup. This can happen as blocks can be fed to the enclave
-		// multiple times. We don't update the state and move on.
-		if len(rollupsInBlock) == 1 && bytes.Equal(rollupsInBlock[0].Header.Hash().Bytes(), genesisRollup.Hash().Bytes()) {
-			return nil, errIsGenesisRollupInBlock
+		// If we haven't stored the genesis rollup and there are no rollups in this block, it cannot be the L2
+		// genesis block.
+		if len(rollupsInBlock) == 0 {
+			return &preGenesis, nil
 		}
-		// We return - we do not need to handle the genesis rollup, since we already have.
-		return nil, nil //nolint:nilnil
+
+		// If we haven't stored the genesis rollup before and this block contains rollups, it must be the L2 genesis
+		// block.
+		return &genesis, nil
 	}
 
-	// The incoming block holds the genesis rollup. Calculate and return the new block state.
-	// todo change this to an hardcoded hash on testnet/mainnet
-	rc.logger.Info("Found genesis rollup", "l1Height", block.NumberU64(), "l1Hash", block.Hash())
+	// We check if we're reprocessing the genesis block.
+	if len(rollupsInBlock) == 1 && bytes.Equal(rollupsInBlock[0].Header.Hash().Bytes(), genesisRollup.Hash().Bytes()) {
+		return &genesis, nil
+	}
+
+	return &postGenesis, nil
+}
+
+// Checks if we're processing the genesis rollup block for the first time, and updates the state if we are.
+func (rc *RollupChain) handleGenesisBlock(block *types.Block, rollupsInBlock []*core.Rollup) (*common.L2RootHash, error) {
+	genesisRollup, err := rc.storage.FetchGenesisRollup()
+	if err != nil && !errors.Is(err, errutil.ErrNotFound) {
+		return nil, fmt.Errorf("could not retrieve genesis rollup. Cause: %w", err)
+	}
+
+	// Check if we've processed the genesis rollup before.
+	if err == nil {
```

Comment:
I don't understand how we can hit this path if we never read the same L1 block twice?